src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1152,100 +868,93 @@ ActivateNodeList(List *nodeList)
 	 */
 	EnsureSuperUser();
 
-	/* take an exclusive lock on pg_dist_node to serialize pg_dist_node changes */
+	/*
+	 * take an exclusive lock on pg_dist_node to serialize pg_dist_node
+	 * changes
+	 */
 	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
 
+	/*
+	 * error if there is concurrent change to node table before acquiring
+	 * the lock
+	 */
+	ErrorIfAnyNodeNotExist(context->allActivatedNodeList);
 
-	List *nodeToSyncMetadata = NIL;
-	WorkerNode *node = NULL;
-	foreach_ptr(node, nodeList)
-	{
-		/*
-		 * First, locally mark the node is active, if everything goes well,
-		 * we are going to sync this information to all the metadata nodes.
-		 */
-		WorkerNode *workerNode =
-			FindWorkerNodeAnyCluster(node->workerName, node->workerPort);
-		if (workerNode == NULL)
-		{
-			ereport(ERROR, (errmsg("node at \"%s:%u\" does not exist", node->workerName,
-								   node->workerPort)));
-		}
-
-		/* both nodes should be the same */
-		Assert(workerNode->nodeId == node->nodeId);
-
-		/*
-		 * Delete existing reference and replicated table placements on the
-		 * given groupId if the group has been disabled earlier (e.g., isActive
-		 * set to false).
-		 *
-		 * Sync the metadata changes to all existing metadata nodes irrespective
-		 * of the current nodes' metadata sync state. We expect all nodes up
-		 * and running when another node is activated.
-		 */
-		if (!workerNode->isActive && NodeIsPrimary(workerNode))
-		{
-			bool localOnly = false;
-			DeleteAllReplicatedTablePlacementsFromNodeGroup(workerNode->groupId,
-															localOnly);
-		}
-
-		workerNode =
-			SetWorkerColumnLocalOnly(workerNode, Anum_pg_dist_node_isactive,
-									 BoolGetDatum(true));
+	/*
+	 * delete existing reference and replicated table placements on the
+	 * given groupId if the group has been disabled earlier (e.g., isActive
+	 * set to false).
+	 */
+	SendDeletionCommandsForReplicatedTablePlacementsInCoordinatedTransaction(context);
```

Comment:
I think if we are going to make this non-transactional, we should switch the order. Even if a row deletion fails remotely, the local row should stay so that we should be able to re-try.

In other words: 


```C
			DeleteShardPlacementRow(placement->placementId);

			if (!localOnly)
			{
				resetStringInfo(deletePlacementCommand);
				appendStringInfo(deletePlacementCommand,
								 "DELETE FROM pg_catalog.pg_dist_placement "
								 "WHERE placementid = " UINT64_FORMAT,
								 placement->placementId);

				SendCommandToWorkersWithMetadata(deletePlacementCommand->data);
			}
```
should become

```C
			if (!localOnly)
			{
				resetStringInfo(deletePlacementCommand);
				appendStringInfo(deletePlacementCommand,
								 "DELETE FROM pg_catalog.pg_dist_placement "
								 "WHERE placementid = " UINT64_FORMAT,
								 placement->placementId);

				SendCommandToWorkersWithMetadata(deletePlacementCommand->data);
			}

			DeleteShardPlacementRow(placement->placementId);

```

And sure, SendCommandToWorkersWithMetadata will change

Reply:
Created a new function which uses `metadatasyncContext` connections. Divided the old function into pieces to share the code as much as possible.